Add Xiaomi miio Alarm Control Panel

starkillerOG opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add new Xiaomi miio Alarm Control Panel integration.
The requered new version of python-miio has been released, so this is now ready for merging!

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

Config Flow Support!

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12167

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

starkillerOG committed:
Add Xiaomi miio Alarm Control Panel

starkillerOG committed:
black formatting

starkillerOG committed:
fix lint issues

starkillerOG committed:
fix issort

starkillerOG committed:
black formatting

starkillerOG committed:
Add Xiaomi miio aqara gateway alarm

rytilahti commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,201 @@
+Support for Xiomi Gateway alarm control panels.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/
```
Is this boilerplate still in the integration templates? I thought that this is not anymore wanted..

starkillerOG replied:
You are right, I changed it in the latest commit

rytilahti reviewed (commented):
This looks fine to me based on a brief read-through. Looks like aot of the code is shared with many other xiaomi_miio platforms, so that is something that could use some cleanup in the future to avoid code duplication.

starkillerOG committed:
update style of boilerplate

rytilahti commented:
0.5.0 release is finally out (https://github.com/rytilahti/python-miio/releases/tag/0.5.0), #32850 will bump the version so this PR can also proceed now :-)

starkillerOG committed:
bump python-miio version to 0.5.0

starkillerOG committed:
bump python-miio to v0.5.0.1

starkillerOG commented:
@MartinHjelmare this is ready to be merged now that python-miio v0.5.0.1 has been released.
@rytilahti thank you for all your help with this PR!

starkillerOG commented:
@syssi could you review this PR since you are a code owner?
(@rytilahti has already revieuwd it)

MartinHjelmare reviewed (changes requested):
This integration needs to be refactored before we can accept any new features or platforms. We don't allow platform config any longer. See ADR 7:

https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md

This is not a new integration but we don't allow new platforms for an existing integration with platform config.

starkillerOG commented:
@rytilahti It seems like all Xiaomi miio integrations need to be adapted to this new configuration structure. Do you want to do this, or shall I attempt to do it?

rytilahti commented:
@starkillerOG I'm currently occupied with other things, so I don't have much time to do any larger contributions on that front, so feel free to work on it if you like. Some of the things related to a long-needed cleanup are listed here: https://github.com/home-assistant/core/pull/33736#pullrequestreview-388173535

starkillerOG committed:
Centralize configuration

starkillerOG committed:
Centralize configuration

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -1 +1,81 @@
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
```
We recently accepted ADR10 which makes this legacy already.

See: https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md

Please refactor to use a config flow and config entries instead.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

starkillerOG replied:
well at least I made a start to centrilize the configuration, I will continue tomorrow and try to update it to the config flow.

The Xiaomi_aqara platform uses this same way of setting up the platform so that integration will probably also need refactoring.

@MartinHjelmare Could the HomeAssistant Dev Docs be updated to reflect these decisions, I got my info from:
https://developers.home-assistant.io/docs/creating_component_generic_discovery/
Under "Loading platforms when configured via configuration.yaml"
The example mentioned there is also depricated in that case:
https://github.com/home-assistant/example-custom-config/tree/master/custom_components/example_load_platform/

MartinHjelmare replied:
Some integrations are still allowed to use config yaml, so we still need that part of the docs. We should probably make it clear when that is an option, though.

starkillerOG committed:
fix overwrites

starkillerOG committed:
Implement Xiaomi Miio Config Flow

starkillerOG committed:
Implement Xiaomi Miio Config Flow

starkillerOG committed:
Implement Xiaomi Miio Config Flow

starkillerOG committed:
Implement Xiaomi Miio Config Flow

starkillerOG committed:
Implement Xiaomi Miio Config Flow

starkillerOG committed:
Implement Xiaomi Miio Config Flow

starkillerOG committed:
Implement Xiaomi Miio Config Flow

starkillerOG committed:
fix overwrites

starkillerOG committed:
remove unused imports

starkillerOG committed:
remove unused imports

starkillerOG committed:
remove unused imports

starkillerOG commented:
@MartinHjelmare after a lot of work I finally got the Config Flow implemented and working for the Xiaomii Miio domain.
I tested it locally and it works.

I implemented a 2 step config flow:
1) the user is asked to select which device to connect to (currently the only option is the Xiaomi Gateway, but since the Xiaomi_miio integration also supports lots of other devices that are currently still configured through Yaml, this allows the other devices such as the robot vacuums etc. to be placed here in this list for future PRs)
2) when the Xiaomi Gateway is selected the user is then prompted to fill in the following:
- Gateway IP
- Gateway API Token
- Optional name of the gateway (a default is set)

after that the xiaomi gateway is configured and a xiaomi gateway alarm entity is made.

Can this PR now finnally be accepted?
If so I will update the docs and then finnaly be done with this PR...

starkillerOG committed:
fix issort

starkillerOG committed:
fix issort

starkillerOG committed:
fix pyupgrade

starkillerOG committed:
fix pyupgrade

starkillerOG committed:
fix faulty unused_import

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+    @property
+    def should_poll(self):
+        """Poll the miio device."""
+        return True
```
This is the default. We can remove the property.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+        self._gateway = gateway_device
+        self._name = gateway_name
+        self._skip_update = False
+        self._device_id = device_id
```
What kind of id is the device id?

starkillerOG replied:
that is the unique_id of the gateway device, this is later used in the `device_info` property to define the identifiers to link it to the gateway device.
I will rename to make it clearer

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+    async def _try_command(self, mask_error, func, *args, **kwargs):
+        """Call a device command handling error messages."""
+        try:
+            result = await self.hass.async_add_job(partial(func, *args, **kwargs))
```
Please use `hass.async_add_executor_job` instead of `hass.async_add_job`. The latter is legacy for general use.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+        try:
+            result = await self.hass.async_add_job(partial(func, *args, **kwargs))
+
+            _LOGGER.info("Response received from miio device: %s", result)
```
Debug.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+
+            _LOGGER.info("Response received from miio device: %s", result)
+
+            return result == XIAOMI_SUCCESS
```
Nothing is checking this return value.

starkillerOG replied:
True, I just liked the function returning True of False if it is succesfull or not, I could make it just return nothing, is that better?

MartinHjelmare replied:
Yeah, we should return values when needed.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+
+    async def async_alarm_arm_away(self, code=None):
+        """Turn on."""
+        await self._try_command("Turning the alarm on failed.", self._gateway.alarm.on)
```
We don't end logging messages with period.

MartinHjelmare replied:
There's no formatting argument specifier in the log message string. But we're passing an argument in the log call, ie the exception argument.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+
+    async def async_update(self):
+        """Fetch state from the device."""
+        # On state change some devices doesn't provide the new
```
don't

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+                self._state = None
+
+            _LOGGER.debug("State value: %s", self._state)
+        except DeviceException as ex:
```
Please only wrap the line(s) that can raise inside the try... except block. Move the `except` up and return after logging the error. Then we can outdent below that.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+        if user_input is not None:
+            # Check which device needs to be connected.
+            if user_input.get(CONF_GATEWAY):
+                return self.async_show_form(
```
When we want to move to the next step, await and return that step, don't call show form.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+        errors = {}
+        host = user_input.get(CONF_HOST)
+        token = user_input.get(CONF_TOKEN)
+        if user_input is not None and host is not None and token is not None:
```
Host and token are required in the input if the input is not `None`.

starkillerOG replied:
does the input not include the data from the previous step (in this case the boolean "gateway")?

starkillerOG replied:
have figured it out, is does not in the way I now call the function.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+        token = user_input.get(CONF_TOKEN)
+        if user_input is not None and host is not None and token is not None:
+            # Try to connect to a Xiaomi Gateway.
+            _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
```
Debug.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+
+            try:
+                self._gateway = gateway.Gateway(host, token)
+                self._gateway_info = self._gateway.info()
```
I/O?

starkillerOG replied:
yes this is I/O

starkillerOG replied:
I split up async_setup_entry to allow for easy addition of the other Xiaomi Miio devices such as robot vacuums.
I also centrilized the connect logic in the gateway.py file in the ConnectXiaomiGateway class.
And I made this connection async to not block hass.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+                    self._gateway_info.hardware_version,
+                )
+
+                unique_id = "{}-{}-gateway".format(
```
Please use f-strings instead of str.format.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+                    self._gateway_info.model, self._gateway_info.mac_address
+                )
+                await self.async_set_unique_id(unique_id)
+                return self.async_create_entry(
```
We should add a call to `self._abort_if_unique_id_configured` here.

MartinHjelmare reviewed (commented):
The model of the config flow sounds good.

The config flow requires 100% test coverage. The scaffolding script can help you get started with the tests.

starkillerOG committed:
typo

starkillerOG committed:
typo

starkillerOG commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -1 +1,79 @@
+
+async def async_setup(hass: core.HomeAssistant, config: dict):
+    """Set up the Xiaomi Miio component."""
+    conf = config.get(DOMAIN)
```
I think this function is called after the Config Flow has finisched and the config includes the user_input of the config flow

MartinHjelmare replied:
No. The config passed here is the config yaml.

starkillerOG replied:
alright, what schould I then do in the async_setup function?
Before I did not have that function at all, but then it would give me an error during setup that that function is missing and the integration would not be setup correctly.

Schould I just have an empty function that imediatly returns True?

MartinHjelmare replied:
Yes, just `return True`. :+1:

starkillerOG commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -1 +1,79 @@
+    token = entry.data.get(CONF_TOKEN)
+    name = entry.title
+    gateway_id = entry.data.get("gateway_id")
+    _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
```
The token is always 32 characters long, we only show the first 5 characters in the log, I thinks that schould be enough obfuscated

starkillerOG commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -1 +1,79 @@
+    # Connect to gateway
+    try:
+        gateway_device = gateway.Gateway(host, token)
+        gateway_info = gateway_device.info()
```
What do you mean with I/O?
This is the place where the gateway_device class is setup, that class connects to the device during its initilization and that class contains all functions to communicate (get/set) with the device.

starkillerOG replied:
The gateway_info is only a class that contains information, it does not get or set anything, it is only needed to get the model number, firmware version etc.

MartinHjelmare replied:
Input/Output. Eg making a request over the network, reading or writing to disk etc.

starkillerOG replied:
so the `gateway_device = gateway.Gateway(host, token)` is I/O since it makes the connection to the device.
The `gateway_info = gateway_device.info()` is also I/O since it makes a request to get the gateway info

MartinHjelmare replied:
Ok, then we need to schedule these calls on the executor thread pool with `hass.async_add_executor_job`.

Try to group both calls into one function that stores or returns the variables. Maybe consider wrapping this into another class that has a setup function that sets the gateway and info as attributes.

starkillerOG replied:
I split up async_setup_entry to allow for easy addition of the other Xiaomi Miio devices such as robot vacuums.
I also centrilized the connect logic in the gateway.py file in the ConnectXiaomiGateway class.
And I made this connection async to not block hass.

starkillerOG committed:
process revieuw requests

starkillerOG commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+
+XIAOMI_STATE_ARMED_VALUE = "on"
+XIAOMI_STATE_DISARMED_VALUE = "off"
+XIAOMI_STATE_ARMING_VALUE = "oning"
```
That is the 'raw' value that the gateway returns if it is turning on the alarm, it will blink its light red during a settable amount of time (like 15 seconds) while playing a sound at the beginning and end. This period allows you to leave the house before the alarm is activated.

'oning' is translated to STATE_ALARM_ARMING of HomeAssistant on line 145-146

MartinHjelmare replied:
Ok.

starkillerOG commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,162 @@
+        self._name = gateway_name
+        self._skip_update = False
+        self._device_id = device_id
+        self._unique_id = f"{model}-{mac_address}-alarm"
```
you are right, there is only 1 per gateway

starkillerOG committed:
process revieuw comments

starkillerOG committed:
cleanup async_setup

starkillerOG committed:
Do not return value if not used

starkillerOG committed:
process revieuw comments

starkillerOG committed:
Create gateway.py

starkillerOG committed:
improve Xiaomi Miio Config Flow

starkillerOG committed:
Split up async_setup_entry & async connect

Split up async_setup_entry to allow for easy addition of the other Xiaomi Miio devices such as robot vacuums.
Centrilize connect logic in the gateway.py file in the ConnectXiaomiGateway class.
Also make this connect async to not block hass.

starkillerOG committed:
remove unused imports

starkillerOG committed:
black formatting

starkillerOG committed:
black formatting

starkillerOG committed:
black formatting

starkillerOG committed:
add missing docstring

starkillerOG commented:
@MartinHjelmare I think I did everything you sugested in the revieuw, I think the code is now good. Only the tests are still missing.

Could you check if the code is now alright?

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -1 +1,71 @@
+
+    # Connect to gateway
+    connect_gateway_class = ConnectXiaomiGateway()
+    await hass.async_add_job(connect_gateway_class.async_connect_gateway(host, token))
```
We should check the return value here, right?

starkillerOG replied:
I did "check the return value", by looking at `if gateway_info is None`, but now I removed that and just derectly check the return value

MartinHjelmare commented on homeassistant/components/xiaomi_miio/alarm_control_panel.py:
```diff
@@ -0,0 +1,150 @@
+    async def async_update(self):
+        """Fetch state from the device."""
+        try:
+            state = await self.hass.async_add_job(self._gateway.alarm.status)
```
Replace all `hass.async_add_job` with either `hass.async_create_task` or `hass.async_add_executor_job` depending on the target. If it's a function that should run in the executor thread pool, use `hass.async_add_executor_job`.

MartinHjelmare replied:
The gateway here is the wrapped gateway from the gateway module, right? We need to adjust the attribute access, right?

starkillerOG replied:
I replaced all hass.async_add_job.
The gateway here is **not** the wrapped gateway from the gateway.py file in HomeAssistant.
The gateway here is the gateway class of the python miio module (v0.5.0.1)

starkillerOG replied:
So I do not think we need to adjust the attribute access

MartinHjelmare commented on homeassistant/components/xiaomi_miio/config_flow.py:
```diff
@@ -0,0 +1,82 @@
+        errors = {}
+        if user_input is not None:
+            # Check which device needs to be connected.
+            if user_input.get(CONF_GATEWAY):
```
Please go through all the code and remove dict.get where we don't need it. There's a default for this item in the config schema.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/gateway.py:
```diff
@@ -0,0 +1,44 @@
+        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+        try:
+            self._gateway_device = gateway.Gateway(host, token)
+            self._gateway_info = self._gateway_device.info()
```
If this does I/O we need to schedule it on the executor thread pool with `hass.async_add_executor_job`.

starkillerOG replied:
I call this function in 2 places:
- in the __init__.py file in the async_setup_gateway_entry, there I have the `hass `class so no problem
- in the config_flow.py file, there I do not have the `hass` class, how can I then call hass.async_add_executor_job?

What I did before was call the function of the gateway class with `hass.async_add_job` which schould have been `hass.async_add_executor_job`, in the __init__.py file, but I just removed that because of your remark about it above in this revieuw.

How schould I resolve this?

MartinHjelmare replied:
In the config flow we have access to `self.hass`.

MartinHjelmare replied:
Is it only the `info` call that does I/O or also the instantiation of `Gateway`?

starkillerOG replied:
I just checked it and indeed only the info call does I/O, the instantation of Gateway only initializes the class and all of its (many) sub classes that contain all functions and info variables needed to communicate with the device, but nothing is actually sent to the device.

Handshaking, retrieving communication tokens, ids etc. is done at the first command that is sent, so in this case that is indeed at the info call that actually communicates with the device.

starkillerOG replied:
Therefore I will only schedule the info call on the executor thread pool.

starkillerOG committed:
process revieuw comments

starkillerOG committed:
process revieuw comments

starkillerOG committed:
use key acces instead of .get

starkillerOG committed:
Run I/O in executor thread pool

starkillerOG committed:
schedule I/O in executor thread pool

starkillerOG committed:
schedule I/O in executor thread pool

starkillerOG commented:
@MartinHjelmare tested it and everthing works, I think everything is done right?

MartinHjelmare commented on homeassistant/components/xiaomi_miio/gateway.py:
```diff
@@ -0,0 +1,46 @@
+class ConnectXiaomiGateway:
+    """Class to async connect to a Xiaomi Gateway."""
+
+    def __init__(self):
```
I'd pass `hass` to the gateway and store `hass` as an instance attribute.

MartinHjelmare replied:
```suggestion
    def __init__(self, hass):
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/gateway.py:
```diff
@@ -0,0 +1,46 @@
+        """Return the class containing gateway info."""
+        return self._gateway_info
+
+    async def async_connect_gateway(self, hass, host, token):
```
Then we don't need to pass `hass` here.

MartinHjelmare commented:
Have we tested setting up this integration both via config flow and by adding another old platform config yaml at the same time?

starkillerOG committed:
Update homeassistant/components/xiaomi_miio/gateway.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
pass hass at initialize

starkillerOG committed:
pass hass at init

starkillerOG committed:
pass hass at init

starkillerOG commented:
> 
> 
> Have we tested setting up this integration both via config flow and by adding another old platform config yaml at the same time?
> 
> We need to make sure we know what the compatibility status is before merging.

I don't own any other Xiaomi Miio devices with wich I could truly test is, but what I did is add a xiaomi miio switch to my configuration:

```
switch:
  - platform: xiaomi_miio
    host: 192.168.1.99
    token: token
```

with a IP adress that is not connected to my network and the same token as I use for my gateway.

I get the following in the log:
```
ERROR (SyncWorker_9) [miio.miioprotocol] Unable to discover a device at address 192.168.1.99
WARNING (MainThread) [homeassistant.components.switch] Platform xiaomi_miio not ready yet. Retrying in 30 seconds.
ERROR (SyncWorker_0) [miio.miioprotocol] Unable to discover a device at address 192.168.1.99
WARNING (MainThread) [homeassistant.components.switch] Platform xiaomi_miio not ready yet. Retrying in 60 seconds.
ERROR (SyncWorker_19) [miio.miioprotocol] Unable to discover a device at address 192.168.1.99
WARNING (MainThread) [homeassistant.components.switch] Platform xiaomi_miio not ready yet. Retrying in 90 seconds.
```

So it seems like setting up a device though the config will work fine, besides the fact that in my case the device does not physically exists.

MartinHjelmare commented:
Good! @rytilahti do you have time to test too?

starkillerOG committed:
create Config Flow test for Xiaomi Miio

starkillerOG committed:
fix test_config_flow with black and isort

starkillerOG commented:
@MartinHjelmare I also included all tests for the Config Flow.
That took me at least as long as setting up the config flow itself to figure out, but I learend a lot about Mocks and patches :)

I think this PR is now fully done and ready for merging.

Is there anything else I schould do before this can be merged?

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+from asyncio import Future
+from unittest.mock import Mock
+
+from asynctest import patch
```
```suggestion
from asynctest import CoroutineMock, patch
from miio import DeviceException
```

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+    mock_gateway = get_mock_gateway(False)
+
+    with patch(
+        "homeassistant.components.xiaomi_miio.config_flow.ConnectXiaomiGateway",
+        return_value=mock_gateway,
```
```suggestion
        "homeassistant.components.xiaomi_miio.gateway.gateway.Gateway.info",
        side_effect=DeviceException({}),
```

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+    with patch(
+        "homeassistant.components.xiaomi_miio.config_flow.ConnectXiaomiGateway",
+        return_value=mock_gateway,
+    ), patch(
+        "homeassistant.components.xiaomi_miio.ConnectXiaomiGateway",
+        return_value=mock_gateway,
```
```suggestion
```

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+    mock_gateway = get_mock_gateway()
+
+    with patch(
+        "homeassistant.components.xiaomi_miio.config_flow.ConnectXiaomiGateway",
+        return_value=mock_gateway,
```
```suggestion
        "homeassistant.components.xiaomi_miio.gateway.gateway.Gateway.info",
        return_value=mock_info,
```

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+    with patch(
+        "homeassistant.components.xiaomi_miio.config_flow.ConnectXiaomiGateway",
+        return_value=mock_gateway,
+    ), patch(
+        "homeassistant.components.xiaomi_miio.ConnectXiaomiGateway",
+        return_value=mock_gateway,
```
```suggestion
```

MartinHjelmare replied:
Instead we need to patch `async_setup_entry` in the integration module to avoid setting up the integration when creating the entry. See eg:

https://github.com/home-assistant/core/blob/5f775c14278074a75464b538e69a66813a31e549/tests/components/hue/test_config_flow.py#L19-L23

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+TEST_FIRMWARE_VERSION = "1.2.3_456"
+
+
+def get_mock_gateway(
+    connect_succes=True,
+    model=TEST_MODEL,
+    mac_address=TEST_MAC,
+    hardware_version=TEST_HARDWARE_VERSION,
+    firmware_version=TEST_FIRMWARE_VERSION,
+):
+    """Return a mock gateway."""
+    mock_gateway = Mock()
+    mock_gateway.async_connect_gateway.return_value = Future()
+    mock_gateway.async_connect_gateway.return_value.set_result(connect_succes)
+    if connect_succes:
+        mock_gateway.gateway_info.model = model
+        mock_gateway.gateway_info.mac_address = mac_address
+        mock_gateway.gateway_info.hardware_version = hardware_version
+        mock_gateway.gateway_info.firmware_version = firmware_version
+    else:
+        mock_gateway.gateway_info = None
+
+    return mock_gateway
```
```suggestion
def get_mock_info(
    model=TEST_MODEL,
    mac_address=TEST_MAC,
    hardware_version=TEST_HARDWARE_VERSION,
    firmware_version=TEST_FIRMWARE_VERSION,
):
    """Return a mock gateway info instance."""
    gateway_info = Mock()
    gateway_info.model = model
    gateway_info.mac_address = mac_address
    gateway_info.hardware_version = hardware_version
    gateway_info.firmware_version = firmware_version

    return gateway_info
```

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+    assert result["step_id"] == "gateway"
+    assert result["errors"] == {}
+
+    mock_gateway = get_mock_gateway()
```
```suggestion
    mock_info = get_mock_info()
```

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,138 @@
+"""Test the Xiaomi Miio config flow."""
+from asyncio import Future
```
```suggestion
```

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Update tests/components/xiaomi_miio/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
fix test_config_flow

starkillerOG commented:
@MartinHjelmare Thank you very much for improving the config flow tests.
I have commited all your changes and made a small fix.

Is it now correct?

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,126 @@
+    assert result["errors"] == {"base": "connect_error"}
+
+
+async def test_config_flow_gateway_succes(hass):
```
success

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,126 @@
+
+
+async def test_config_flow_step_user_no_device(hass):
+    """Test config flow ."""
```
Test user step with no device selected.

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,126 @@
+
+
+async def test_config_flow_step_gateway_connect_error(hass):
+    """Test config flow ."""
```
Test gateway connection error.

MartinHjelmare commented on tests/components/xiaomi_miio/test_config_flow.py:
```diff
@@ -0,0 +1,126 @@
+
+
+async def test_config_flow_gateway_succes(hass):
+    """Test config flow ."""
```
Test a successful flow.

MartinHjelmare commented:
Please also include the config flow module in coverage calculation by explicitly excluding all the other xiaomi_miio modules in `.coveragerc`.

rytilahti commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -1 +1,69 @@
+
+    hass.data[DOMAIN][entry.entry_id] = gateway.gateway_device
+
+    gateway_model = f"{gateway_info.model}-{gateway_info.hardware_version}"
```
Why combine both model and hardware version here? Does model not carry enough information here? This function could be generalized to `async_setup_device_entry` to allow easier porting of other platforms to use the config flow in the future.

starkillerOG replied:
I included the hardware_version since there is no specific device property in which I could place it, see https://developers.home-assistant.io/docs/device_registry_index.

It is nice for users to be able to easily find there hardware_version besides the model, because if a user is experiancing problems whith commands not working/firmware updates or other problems it is nice to compare hardware versions with other users and see if others with exactly the same device (model and hardware version) experiance the same issues.

rytilahti replied:
I agree that it'd would be useful to have the hardware information available (and it is not just specific for xiaomi_miio), but I would much rather prefer to have it fixed in the core. This will require an architecture issue first, though...

starkillerOG replied:
Well that is definetly for some other PR.
I will leave it as it is and it can be changed later on if someone want to make an architecture issue

rytilahti commented on homeassistant/components/xiaomi_miio/gateway.py:
```diff
@@ -0,0 +1,47 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class ConnectXiaomiGateway:
```
This file and could be generalized to allow any sort of xiaomi_miio devices, which will allow us to greatly remove code duplication from all platforms (making other platforms to use this class should go into a separate PR later on).

I'm not sure what would be a good name for the class, but the properties could be renamed:
* gateway_device => device
* info => info

rytilahti commented on homeassistant/components/xiaomi_miio/gateway.py:
```diff
@@ -0,0 +1,47 @@
+class ConnectXiaomiGateway:
+    """Class to async connect to a Xiaomi Gateway."""
+
+    def __init__(self, hass):
```
API wise it would be better to add `cls` parameter here, and pass the `Gateway` as a parameter to the class. This will make it generic for other platforms to use.

rytilahti reviewed (commented):
A quick review, it would be great to generalize a couple of parts to allow other xiaomi_miio platforms to be ported over in the near future.

I'll try to find my gateway and do some testing over the weekend.

starkillerOG committed:
fix function docstrings

starkillerOG committed:
include xiaomi_miio config flow

starkillerOG commented:
> 
> 
> Please also include the config flow module in coverage calculation by explicitly excluding all the other xiaomi_miio modules in `.coveragerc`.

I made a commit for this, I think these are all the files that need to be excluded, is that correct?

MartinHjelmare commented:
Add `gateway.py` and `__init__.py`.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/.translations/en.json:
```diff
@@ -0,0 +1,29 @@
+{
```
We reworked our translations structure. Please rebase on  dev branch and rename the translations directory `.translations` to `translations`.

Also check the result by running the translation dev script again.

starkillerOG replied:
Is it possible to rebase from the web interface of Github?

MartinHjelmare replied:
Not the default interface I think.

starkillerOG replied:
I have downloaded the github desktop client and I think I did a succesfull rebase

starkillerOG committed:
Update .coveragerc

starkillerOG commented:
@rytilahti I agree that making it more general is a good idea, however that is for some next PR.
I already made much of the Config Flow as generic as I could and I have put a lot of tought and effort into room for the oter platforms later on (like making the first form a selection list to select wich platform to set up)

This PR is already 74 commits long and is already taking over 2 months.
I have put a lot of my hours and effort in it and at this point I am kind of done with it.
The requerements have changed over and over again and what started out as just adding the alarm panel with simple On/Off/State has already grew to a much bigger project.

I will fix the translation stuff that has apperently changed, I will look into that tommorow. But I am not going to make any more structural changes. That will have to be done in a later PR. Maybe I will do it later on, but now I just want this to be done.

@MartinHjelmare I hope you will merge this after the translation fix, otherwise I will put this on hold untill I feel like working on it again.

starkillerOG committed:
rename translations folder

MartinHjelmare commented:
If someone more can test and make sure the old platforms work with this we can merge.

starkillerOG commented:
@MartinHjelmare Thank you very much!
@rytilahti could you test if the other platforms still work with this PR?

rytilahti commented:
Alas, I wasn't able to find the old gateway device, but I'll find some time tomorrow to test this with my other devices.

starkillerOG committed:
add missing title in translations

starkillerOG committed:
add missing title in translations

starkillerOG committed:
undo title in translations

whoops was looking at the old docs, this has been changed in version .109

starkillerOG committed:
undow add title in tanslation

starkillerOG commented:
@rytilahti did you have time to test?

rytilahti commented:
I found my gateway device, so I was able to test it! Looks working fine to me, but I have never used the alarm functionality myself so I'm not sure what to expect. I was able to add it, I can arm and disarm (arming causing its light to turn red, and I'm hearing beeps). Also my smart strip & vacuum are still working as expected, so I think it's all good.

It's better to fix this deprecation warning (and unit tests) before merging this:
```
[homeassistant.components.alarm_control_panel] AlarmControlPanel is deprecated, modify XiaomiGatewayAlarm to extend AlarmControlPanelEntity
```

starkillerOG commented:
@rytilahti thank you for testing and your feedback!
@MartinHjelmare I did not see that deprication warning, is that something new in the latest dev version of HomeAssistant?

I am not sure about the naming. I think it lists as Xiaomi Gateway because of the `self.async_create_entry` function inside the config_flow.py file. But I think other platforms such as the vacuum will use another instance of that function with a diffrent title and diffrent data to set up that platform. But I am not entirely sure.

@MartinHjelmare do you know what is going on with the python checks? It lists over 4000 failed tests!
But I only changed somthing in the translation files and then reverted that same change...

starkillerOG committed:
update AlarmControlPanel to AlarmControlPanelEntity

starkillerOG commented:
@rytilahti if I look at the hue component that has been my guidline for the whole config flow, they also use:

```
            return self.async_create_entry(
                title=bridge.config.name,
                data={
                    "host": bridge.host,
                    "username": bridge.username,
                    CONF_ALLOW_HUE_GROUPS: False,
                },
            )
```

where the title is the name of the bridge, but if you have multiple bridges (which is supported) with diffrent names you would also get multiple titles, not sure how that shows up in the integrations list, but I figured it will work somehow...

(I have 1 hue brigde and that lists as "Hue Bridge 2.0", wondering what would happen if I were to add another bridge with a diffrent name....)

rytilahti commented:
@starkillerOG sounds good to me, I have honestly no deep insight how this is supposed to work, so that was more food for thought kind of note. This PR provides a good starting point to add more devices, so thanks for doing the legwork to make it happen :1st_place_medal:

starkillerOG commented:
@MartinHjelmare I think this is then ready to be merged

starkillerOG commented:
@MartinHjelmare However I am a little conserned about all the test failing, but I can't emagine that it truly is a problem of this PR, since there have not been any significant changes since the last Commit that did pass all the tests.

Do you know what is going on?

MartinHjelmare commented:
We're having issues with CI on dev branch at the moment. Stay tuned. I'll get back here when it's fixed and we can rebase.

Regarding the titles, that should not be a breaking change if we need to change it in the future. If an integration has different types of devices and exits the config flow by creating entries for those devices, each create entry step can set a different (sub)title for that device type. This title will show in the frontend below the main title for the integration.

![config_entry_titles](https://user-images.githubusercontent.com/3181692/80426084-4d0ff080-88e5-11ea-9ec5-b7de49cb2582.png)

starkillerOG commented:
I rebased and now only have 1 unrelated failed test with the python38, something about homekit, but that has nothing to do with this PR.

MartinHjelmare commented:
Yeah, that test has been fixed now too.

Great work!

MartinHjelmare commented:
I forgot to approve. Sorry! :heavy_check_mark:

starkillerOG commented:
Thank you for all your help during this PR!
I am verry glad that it finally succeeded :)
